### PR TITLE
Shepherd support for Ark Backup/Restore + Sync operator objects

### PR DESCRIPTION
**Pull request by acbramley**
**This PR does the following:**
- Rips out all of the Backup logic for creating Backup nodes (still need to figure out what to do with existing Backup nodes and configuration)
- Adds direct communication with Openshift in order to issue Backup and Restore operations, creating the CRs, listing them in custom controllers, etc.
- Adds Backup schedule field to Environment type taxonomy in order to schedule backups for all environments of a type. This means there's 1 config to schedule all production environments to backup nightly (for example). This could be extended to add an override at the environment level etc quite easily.
- Abstracts syncing environments into its own operation. This is no longer a Backup one environment then restore to another since Ark holds a lot more information than just the database and files. Instead we have a Sync CRD, which when created will trigger operator logic to do this for us https://gitlab.adelaide.edu.au/web-team/shepherd-environment-sync
- Environments now have a flag to say "Backup my volumes", this tells ark to take a snapshot of the shared PVC when backing up an environment.
- Adds labels to all objects that relate to an environment, this is what Ark uses to gather the objects it needs to backup.
- Deletes all the queue worker stuff around backups, Ark does this for free for us.
- A bunch of minor refactoring and tidy ups, protecting against deleting sites etc.

Relies on https://github.com/universityofadelaide/openshift-client/pull/30

**Benefits of this:**
- Drupal no longer holds the state of the backups, meaning we have much more flexibility in exposing information about the backups/restores. E.g on the restore form you can now only selected Backups which have completed successfully.
- We can track the progress of a backup/restore through the UI (the Phase)
- Abstracts logic for backing up/restoring/syncing things into more manageable chunks (no more scary bash scripts in Drupal config!!!)

**This PR doesn't do the following:**
Testing...again with the lack of testing infrastructure set up to allow mocking of clients etc it's very difficult to test this functionality and I simply ran out of time.

### CODE REVIEW

File: web/modules/custom/shepherd/shp_backup/src/Controller/BackupList.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      '#rows' => [],
+      '#empty' => $this->t('No backups for this site yet.'),
+    ];
+    /** @var \UniversityOfAdelaide\OpenShift\Objects\Backups\BackupList $backup_list */
+    if (!$backup_list = $this->orchestrationProvider->getBackupsForSite($node->id())) {
+      return $table;
+    }
+    foreach ($backup_list->getBackupsByStartTime() as $backup) {
+      $environment = $this->nodeStorage->load($backup->getLabel('environment_id'));
+      // Protect against environments that have been deleted.
```
Comment (pingers) [inline web/modules/custom/shepherd/shp_backup/src/Controller/BackupList.php]: I'm not sure about this part given environments are getting deleted at the moment... we'll see :smile:

Comment (acbramley) [reply web/modules/custom/shepherd/shp_backup/src/Controller/BackupList.php]: What do you mean? This is protecting against that haha

**Comment by singularo** [comment]
Starting to look at this, a couple of questions @acbramley Not sure if you can help, Is there extra permission at the openshift/kubernetes level needed to be able to run the setup commands like `kubectl apply -f examples/common/00-prereqs.yaml` ? (the docs say config/common/00-prereqs.yaml, but the appears to have moved).

There is no mention of any extra requirements in the INSTALL.md changes, but trying to apply the yml gives "Error from server (Forbidden): error when retrieving current configuration of:... and a bunch more"

**Comment by singularo** [comment]
Ahh, so it requires:
```oc login -u system:admin
oc adm policy add-cluster-role-to-user cluster-admin admin```

**Comment by singularo** [comment]
Hmm so i'm stuck with this atm:

```
An error occurred: restores.ark.heptio.com is forbidden: User "system:serviceaccount:myproject:shepherd" cannot list restores.ark.heptio.com in the namespace "heptio-ark": User "system:serviceaccount:myproject:shepherd" cannot list restores.ark.heptio.com in project "heptio-ark": [clusterrole.rbac.authorization.k8s.io "restores.ark.heptio.com" not found, clusterrole.rbac.authorization.k8s.io "backups.ark.heptio.com" not found]
```

I did this:
```
oc adm policy add-cluster-role-to-user backups.ark.heptio.com system:serviceaccount:myproject:shepherd
oc adm policy add-cluster-role-to-user restores.ark.heptio.com system:serviceaccount:myproject:shepherd
```

Did you use oc adm or a kubectl command to allow the service account access to the other namespace @acbramley  or am I barking up the wrong tree there?

**Comment by acbramley** [comment]
@singularo I have documentation on how to grant the shepherd service account access to these resources, I'm not sure which role you're trying to add there. Try following the docs under INSTALL.md https://github.com/universityofadelaide/shepherd/pull/117/files#diff-7d442b7eb49f5fc377f51e74b291cfc1R119

**Comment by singularo** [comment]
Ahh missed that, sorry.

File: INSTALL.md [hunk 2]
```diff
@@ -99,6 +99,44 @@ oc create secret generic privileged-db-password --from-literal=DATABASE_PASSWORD
...
+
+Add a ClusterRole and Binding to the shepherd service account for managing the Ark objects:
+```bash
+oc create clusterrole ark-backups --verb=get,list,create,update,delete --resource=backups,restores,schedules
+oc adm policy add-cluster-role-to-user ark-backups --serviceaccount=shepherd
+```
+
+With the Ark client/server setup, we need to install our plugin:
+```bash
+ark plugin add previousnext/shepherd-ark-mysql-plugins
```
Comment (singularo) [inline INSTALL.md]: How does this bit work, where does the plugin actually come from?

Comment (singularo) [reply INSTALL.md]: Never mind, I see it in hub.docker.com. Hmm we probably need the source for the plugins for future.

Comment (pingers) [reply INSTALL.md]: I think we're supposed to have access - @acbramley took me through it during handover... I assumed it was a public repo.

Comment (acbramley) [reply INSTALL.md]: The repo is on your gitlab, it's just published to a public image on our docker hub for now. Feel free to change that.

**Pull request closed without merging**